clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,174 @@
+  // expected-error@#8{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel3>(F3); // 8th instantiation throwing error
+}
+
+template <typename Func>
+void kernel4func(const Func &F4) {
+  constexpr const char *F4_output = __builtin_unique_stable_name(F4);
+  // expected-error@#9{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel4>([]() {}); // 9th instantiation throwing error
+}
+
+template <template <typename> typename Outer, typename Inner>
+struct S {
+  void operator()() const;
+};
+
+template <typename Ty>
+struct T {};
+
+template <typename Func>
+void kernel13_14func(const Func &F) {
+  // expected-error@#10{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel13>(F); // 10th instantiation throwing error
+  // expected-error@#11{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel14>(F); // 11th instantiation throwing error
+  // Using the same functor twice should be fine
+}
+
+template <typename Ty>
+auto func() -> decltype(__builtin_unique_stable_name(Ty::str));
+
+struct Derp {
+  static constexpr const char str[] = "derp derp derp";
+};
+
+template <typename T>
+static constexpr const char *output1 = __builtin_unique_stable_name(T);
+
+#define MACRO12()                \
+  auto l12 = []() { return 1; }; \
+  constexpr const char *l1_output = __builtin_unique_stable_name(l12);
+
+int main() {
+
+  // kernel0
+  // kernel_single_task<class kernel0>(func<Derp>());
+
+  // kernel1 Evaluate the builtin in a constant context, then call the kernel with that same type
+  auto l1 = []() {};
+  constexpr const char *l1_output = __builtin_unique_stable_name(l1);
+  kernel_single_task<class kernel1>(
+      [=]() { l1(); });
+
+  // kernel2 Call to the builtin on a lambda, then using the lambda for a kernel name
+  auto l2 = []() { return 1; };
+  constexpr const char *l2_output = __builtin_unique_stable_name(l2); // expected-note {{'__builtin_unique_stable_name' evaluated here}}
+  // expected-error@#1{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel2>(l2); // 1st instantiation throwing error
+
+  // kernel3
+  // The current function is named with a lambda (that is, takes a lambda as a template parameter).  Call the builtin (perhaps on the current function?), then pass it to a kernel.
```

Comment:
@schittir -- In test files, it's fine for expected-whatever lines to run over the 80-col limit, but you should wrap "typical" comments to the 80 col limits.